File: backend/src/main/java/br/com/centralmax/maxhub/activity/ActivityFeedService.java
package br.com.centralmax.maxhub.activity;

import br.com.centralmax.maxhub.activity.dto.ActivityFeedResponse;
import br.com.centralmax.maxhub.common.response.PageResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ActivityFeedService {

    private final ActivityFeedRepository repository;

    @Async
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void log(UUID userId, String userName, String actionType,
                    String entityType, UUID entityId, String entityLabel, String details) {
        ActivityFeedEntry entry = ActivityFeedEntry.builder()
                .userId(userId)
                .userName(userName)
                .actionType(actionType)
                .entityType(entityType)
                .entityId(entityId)
                .entityLabel(entityLabel)
                .details(details)
                .build();
        repository.save(entry);
    }

    @Transactional(readOnly = true)
    public PageResponse<ActivityFeedResponse> list(int page, int size) {
        return PageResponse.from(
                repository.findAllByOrderByCreatedAtDesc(PageRequest.of(page, size))
                        .map(this::toResponse)
        );
    }

    @Transactional(readOnly = true)
    public java.util.List<ActivityFeedResponse> listRecent(int limit) {
        return repository.findAllByOrderByCreatedAtDesc(PageRequest.of(0, limit))
                .map(this::toResponse)
                .toList();
    }

    private ActivityFeedResponse toResponse(ActivityFeedEntry e) {
        return new ActivityFeedResponse(e.getId(), e.getUserId(), e.getUserName(),
                e.getActionType(), e.getEntityType(), e.getEntityId(),
                e.getEntityLabel(), e.getDetails(), e.getCreatedAt());
    }
}

File: frontend/src/services/activityFeedService.ts
import { api } from './api';

export interface ActivityFeedEntry {
  id: string;
  userId: string;
  userName: string;
  actionType: string;
  entityType: string;
  entityId: string | null;
  entityLabel: string | null;
  details: string | null;
  createdAt: string;
}

export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

export async function listActivityFeed(page = 0, size = 30): Promise<PageResponse<ActivityFeedEntry>> {
  const { data } = await api.get<PageResponse<ActivityFeedEntry>>('/activity-feed', {
    params: { page, size },
  });
  return data;
}

export async function listRecentActivity(limit = 10): Promise<ActivityFeedEntry[]> {
  const { data } = await api.get<ActivityFeedEntry[]>('/activity-feed/recent', {
    params: { limit },
  });
  return data;
}
